mavros_extras/src/plugins/wheel_odometry.cpp

```diff
@@ -0,0 +1,525 @@
+			}
+			// Duplicate 1st wheel if only one is available.
+			// This generalizes odometry computations for 1- and 2-wheels configurations.
+			else if (wheel_radius.size() == 1) {
+				wheel_offset.resize(2);
+				wheel_radius.resize(2);
+				wheel_offset[1].x() = wheel_offset[0].x();
+				wheel_offset[1].y() = wheel_offset[0].y() + 1.f; // make separation non-zero to avoid div-by-zero
+				wheel_radius[1] = wheel_radius[0];
+			}
+
+			// Check for non-zero wheel separation (first two wheels)
+			float separation = std::abs(wheel_offset[1].y() - wheel_offset[0].y());
+			if (separation < 1.e-5f) {
+				odom_mode = OM::NONE;
+				ROS_WARN_NAMED("wo", "WO: Separation between the first two wheels is too small (%f).", separation);
+			}
+
+			// Check for reasonable radiuses
+			for (int i = 0; i < wheel_radius.size(); i++) {
+				if (wheel_radius[i] <= 1.e-5f) {
+					odom_mode = OM::NONE;
+					ROS_WARN_NAMED("wo", "WO: Wheel #%i has incorrect radius (%f).", i, wheel_radius[i]);
+				}
+			}
+		}
+
+		// Advertise RPM-s and distance-s
+		if (diagnostics) {
+			rpm_pub = wo_nh.advertise<mavros_msgs::Float32ArrayStamped>("rpm", 10);
+			dist_pub = wo_nh.advertise<mavros_msgs::Float32ArrayStamped>("distance", 10);
+		}
+
+		// Advertize topics
+		if (odom_mode != OM::NONE) {
+			if (twist_send)
+				twist_pub = wo_nh.advertise<geometry_msgs::TwistWithCovarianceStamped>("velocity", 10);
+			else
+				odom_pub = wo_nh.advertise<nav_msgs::Odometry>("odom", 10);
+		}
+		// No-odometry warning
+		else
+			ROS_WARN_NAMED("wo", "WO: No odometry computations will be performed.");
+
+	}
+
+	Subscriptions get_subscriptions()
+	{
+		return {
+			make_handler(&WheelOdometryPlugin::handle_rpm),
+			make_handler(&WheelOdometryPlugin::handle_wheel_distance)
+		};
+	}
+
+private:
+	ros::NodeHandle wo_nh;
+
+	ros::Publisher rpm_pub;
+	ros::Publisher dist_pub;
+	ros::Publisher odom_pub;
+	ros::Publisher twist_pub;
+	tf2_ros::TransformBroadcaster tfBroadcaster;
+
+	/// @brief Odometry computation modes
+	enum class OM {
+		NONE,	//!< no odometry computation
+		RPM,	//!< use wheel's RPM
+		DIST	//!< use wheel's cumulative distance
+	};
+	OM odom_mode; //!< odometry computation mode
+
+	int count;		//!< requested number of wheels to compute odometry
+	bool diagnostics;	//!< publish wheel's RPM and cumulative distance
+	std::vector<Eigen::Vector2f> wheel_offset; //!< wheel x,y offsets (m,NED)
+	std::vector<float> wheel_radius; //!< wheel radiuses (m)
+
+	bool twist_send;		//!< send geometry_msgs/TwistWithCovarianceStamped instead of nav_msgs/Odometry
+	bool tf_send;			//!< send TF
+	std::string frame_id;		//!< origin frame for topic headers
+	std::string child_frame_id;	//!< body-fixed frame for topic headers
+	std::string tf_frame_id;	//!< origin for TF
+	std::string tf_child_frame_id;	//!< frame for TF and Pose
+	float vel_cov;			//!< wheel velocity measurement error 1-var (m/s)
```

Comment:
Why float? Messages as i remember uses double. Better use double in calculations.

Reply:
I do all computations in `floats`, because I think float precision is enough for wheel odometry. So I wanted to save some resources.

Reply:
Difference in resource usage not worth it. Better use eigen, as it try to use vector math coprocessors (sse/avx/neon etc).